### ISSUE TITLE
Custom fields in config files ?

### ISSUE THREAD

**Issue by dufguix** [serverpod/serverpod#633]
Hello,

Is it possible to use development.yaml, production.yaml, staging.yaml to add custom config ?
I see it's possible for passwords.yaml.

I think about smtp config for now.

Have a good day.

**Comment by vlidholt** [comment]
That would be a nice addition! I guess it's technically possible to use the passwords file for this until this feature is implemented.

**Comment by dufguix** [comment]
Yes for sure. I'll create a passwords.yaml template without passwords to be able to git push.

**Comment by vlidholt** [comment]
An option is also to do the configuration in code, rather than having it in a config file. Or just load a different config file depending on you run mode. You can access the run mode from `Serverpod.instance!.runmode`. It can be set to one of: `ServerpodRunMode.development`, `ServerpodRunMode.staging`, or `ServerpodRunMode.production`.

**Comment by dufguix** [comment]
Thanks. Good to know `Serverpod.instance!.runmode`.

**Comment by egyleader** [comment]
I can see this is already implemented in serverpod 1.1.0 , I think this issue should be closed 
<img width="282" alt="image" src="https://user-images.githubusercontent.com/31688162/235862311-8df01de7-dfdb-4163-bb77-9b9407ec5973.png">

**Comment by migalv** [comment]
It seems it's not yet implemented if I'm not mistaken. I think this would be very cool to have.

**Comment by SandPod** [comment]
I will update the title of this issue to cover some additional requirements.

## Description
To support advanced use cases and plugins, it should be possible for developers to extend the `<runMode>.yaml` configuration file with their own custom configuration options. These custom configurations should integrate seamlessly with Serverpod’s existing configuration system.

## Requirements

- Developers should be able to register additional configuration fields under custom namespaces.
- It must be possible to specify validation rules and metadata for each configuration field:
  - Whether the field is required or optional.
  - Default values for optional fields.
- Serverpod should produce a **typed configuration object** for the custom configuration:
  - This type can either be supplied by the user at registration or generated by Serverpod.
- The resolution order for custom configuration values must match the standard Serverpod configuration system:
  - Values from environment variables should override values from the configuration file (`<runMode>.yaml`).
- Clear error messages when required fields are missing or have invalid types.
